unixfs.ts

```diff
@@ -0,0 +1,488 @@
+ * describes UinxFS File representation and not IPFS file representation which
+ * is why `RawNode` variant is not part of it.
+ */
+export type FileLayout =
+  | SimpleFileLayout
+  | AdvancedFileLayout
+
+
+/**
+ * Logacal representation of a directory that fits single block.
+ */
+export interface FlatDirectoryLayout extends PBNode {
+  Data: ByteView<{
+    Type: DataType.Directory
+    /**
+     * Directories MUST have `filesize` set to `0`.
+     */
+    filesize: 0
+    /**
+     * If omitted to be interpreted as default `0755`. It is RECOMMENDED
+     * to omit if mode matches default.
+     */
+    mode?: Mode
+    mtime?: UnixTime
+  }>
+  /**
+   * Links are directory entries.
+   */
+  Links: NamedLink<FileLayout|DirectoryLayout>[]
+}
+
+
+/**
+ * Logical representation of directory encoded in multiple blocks (usually when
+ * it contains large number of entries). Such directories are represented via
+ * Hash Array Map Tries (HAMT).
+ * 
+ * @see https://en.wikipedia.org/wiki/Hash_array_mapped_trie
+ */
+export interface AdvancedDirectoryLayout {
+  Data: ByteView<{
+    Type: DataType.HAMTShard,
+    Data: ByteView<Bitfield>
+    /*
+     * HAMT table width (In IPFS it's usually 256)
+     */
+    fanout: uint64,
+    /**
+     * Multihash code for the hashing function used (In IPFS it's [murmur3-64][])
+     *
+     * [murmur3-64]:https://github.com/multiformats/multicodec/blob/master/table.csv#L24
+     */
+    hashType: uint64,
+
+    /**
+     * If omitted to be interpreted as default `0755`. It is RECOMMENDED
+     * to omit if mode matches default.
+     */
+    mode?: Mode
+    mtime?: UnixTime
+  }>
+  Links: NamedLink<FileLayout|DirectoryLayout|DirectoryShard>[]
+}
+
+/**
+ * 
+ */
+export interface DirectoryShard extends PBNode {
+  Data: ByteView<{
+    Type: DataType.HAMTShard,
+    Data: ByteView<Bitfield>
+    /*
+     * HAMT table width (In IPFS it's usually 256)
+     */
+    fanout: uint64,
+    /**
+     * Multihash code for the hashing function used (In IPFS it's [murmur3-64][])
+     *
+     * [murmur3-64]:https://github.com/multiformats/multicodec/blob/master/table.csv#L24
+     */
+    hashType: uint64,
+  }>
+  /**
+   * Either links to other shards or actual directory entries
+   */
+  Links: NamedLink<FileLayout|DirectoryLayout|DirectoryShard>[]
+}
+
+/**
+ * Type for either UnixFS directory representation.
+ */
+export type DirectoryLayout =
+  | FlatDirectoryLayout
+  | AdvancedDirectoryLayout
+
+
+/**
+ * @TODO
+ */
+export type MetadataNode = never
+
+/**
+/**
+ * Logical representation of a [symbolic link][] .
+ *
+ * [symbolic link]:https://en.wikipedia.org/wiki/Symbolic_link
+ */
+ */
+export interface Symlink extends PBNode {
+  Data: ByteView<{
+    Type: DataType.Symlink,
+    /**
+     * UTF-8 encoded path to the symlink target.
+     */
+    Data: ByteView<string>
+    /**
+     * Number of bytes in Data field
+     */
+    filesize: uint64
```

Comment:
Can symlinks have mtime / mode ? They also [seem to have blocksizes](https://explore.ipld.io/#/explore/QmUGzmqt6pcUnghcu68UtkaR8i2r9BGuKQFUNbR1yVTsH8/dave.txt) 

Are they supposed to or is that artifact of normalization no js-ipfs-unixfs side ?

<img width="765" alt="image" src="https://user-images.githubusercontent.com/21236/152038033-3075ed7c-9133-4110-ad4d-d56fb54f5224.png">